src/mono/System.Private.CoreLib/src/System/Reflection/RuntimeParameterInfo.cs

```diff
@@ -131,38 +131,123 @@ internal RuntimeParameterInfo(MethodInfo owner, string? name, Type parameterType
             AttrsImpl = ParameterAttributes.None;
         }
 
-        public override
-        object? DefaultValue
+        private object? GetDefaultValueFromCustomAttributeData()
         {
-            get
+            foreach (CustomAttributeData attributeData in RuntimeCustomAttributeData.GetCustomAttributes(this))
             {
-                if (ClassImpl == typeof(decimal) || ClassImpl == typeof(decimal?))
+                Type attributeType = attributeData.AttributeType;
+                if (attributeType == typeof(DecimalConstantAttribute))
                 {
-                    /* default values for decimals are encoded using a custom attribute */
-                    DecimalConstantAttribute[] attrs = (DecimalConstantAttribute[])GetCustomAttributes(typeof(DecimalConstantAttribute), false);
-                    if (attrs.Length > 0)
-                        return attrs[0].Value;
+                    return GetRawDecimalConstant(attributeData);
                 }
-                else if (ClassImpl == typeof(DateTime) || ClassImpl == typeof(DateTime?))
+                else if (attributeType.IsSubclassOf(typeof(CustomConstantAttribute)))
                 {
-                    /* default values for DateTime are encoded using a custom attribute */
-                    DateTimeConstantAttribute[] attrs = (DateTimeConstantAttribute[])GetCustomAttributes(typeof(DateTimeConstantAttribute), false);
-                    if (attrs.Length > 0)
-                        return attrs[0].Value;
+                    if (attributeType == typeof(DateTimeConstantAttribute))
+                    {
+                        return GetRawDateTimeConstant(attributeData);
+                    }
+                    return GetRawConstant(attributeData);
                 }
-                return DefaultValueImpl;
             }
+            return DBNull.Value;
+        }
+
+        private static decimal GetRawDecimalConstant(CustomAttributeData attr)
```

Comment:
Should this go to shared SPC location?

Reply:
That was my initial thought and I discussed it with @jkotas. 

My understanding is that there is a larger project of moving other pieces of `System.Reflection` code into a sharable component, planned for future (possibly .net8), so I thought of creating a separate issue which would address this and be part of that project (to which I would gladly contribute to) as this PR already grew a bit.

However, if you think moving the code to the shared location should be included here, please let me know.

Reply:
The common piece is less than 100 lines here. My point was that the sharing ceremony with split files, etc. is not worth it for less than 100 lines.